sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1298,6 +1298,103 @@ int gnrc_netif_ipv6_add_prefix(gnrc_netif_t *netif,
 out:
     return res;
 }
+
+#if IS_USED(MODULE_GNRC_NETIF_BUS)
+static bool _has_global_addr(gnrc_netif_t *netif)
+{
+    bool has_global = false;
+
+    gnrc_netif_acquire(netif);
+
+    for (unsigned i = 0; i < CONFIG_GNRC_NETIF_IPV6_ADDRS_NUMOF; i++) {
+        if (ipv6_addr_is_unspecified(&netif->ipv6.addrs[i])) {
+            continue;
+        }
+        if (!ipv6_addr_is_link_local(&netif->ipv6.addrs[i])) {
+            has_global = true;
+            break;
+        }
+    }
+
+    gnrc_netif_release(netif);
+    return has_global;
+}
+
+static void _netif_bus_attach_and_subscribe_addr_valid(gnrc_netif_t *netif,
+                                                       msg_bus_entry_t *sub)
+{
+    msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+    msg_bus_attach(bus, sub);
+    msg_bus_subscribe(sub, GNRC_IPV6_EVENT_ADDR_VALID);
+}
+
+static void _netif_bus_detach(gnrc_netif_t *netif, msg_bus_entry_t *sub)
+{
+    msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
+    msg_bus_detach(bus, sub);
+}
+
+bool gnrc_netif_ipv6_wait_for_global_address(gnrc_netif_t *netif,
+                                             uint32_t timeout_ms)
+{
+    msg_bus_entry_t subs[gnrc_netif_numof()];
+    bool has_global = false;
+
+    if (netif) {
+        if (_has_global_addr(netif)) {
+            return true;
+        }
+
+        _netif_bus_attach_and_subscribe_addr_valid(netif, &subs[0]);
+    } else {
+        unsigned count = 0;
+
+        /* subscribe to all interfaces */
+        while ((netif = gnrc_netif_iter(netif))) {
+            if (_has_global_addr(netif)) {
+                has_global = true;
+            }
+
+            _netif_bus_attach_and_subscribe_addr_valid(netif, &subs[count]);
+            ++count;
```

Comment:
wrt https://github.com/RIOT-OS/RIOT/pull/16824#discussion_r710076990, maybe tying `count` and `netif` somewhat closer together using C-syntax will prevent future breakage:

```suggestion
        /* subscribe to all interfaces */
        for (unsigned count = 0;
             (netif = gnrc_netif_iter(netif));
             count++) {
            if (_has_global_addr(netif)) {
                has_global = true;
            }

            _netif_bus_attach_and_subscribe_addr_valid(netif, &subs[count]);
```